### PR TITLE
Stage #4

### PR DESCRIPTION
**Pull request by DeFiDev**
1. Includes all the latest liquidity functional:
Rpcs:
- createpoolpair
- updatepoolpair
- addpoolliquidity/removepoolliquidity
- poolswap
Rpc listers:
- listpoolpairs
- getpoolpair
- listpoolshares
2. Includes 'government' feature. Rpcs:
- 'setgov' with two (for now) implemented "variables":
	- LP_DAILY_DFI_REWARD
	- LP_SPLITS
- getgov
3. "totalSupply" control -, **not implemented.**

The very latest changes:
- modified "updatetoken" rpc (includes other token's flags) with new auth logic
- modified "minttokens" rpc with new auth logic
- new rpc "updatepoolpair" (changing status, trade commission and even owner)
- all token's flags logic reviewed
- pool trading on/off by pool "status" property

Update:
Fork heights validations unified (as in master) for the rest txs.

### CODE REVIEW

**Review by monstrobishi — CHANGES_REQUESTED** [review]
C++ error in linux build, please fix (See CICD linux build)

File: src/masternodes/govvariables/lp_daily_dfi_reward.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2019 The DeFi Foundation
```
Comment (ShengguangXiao) [inline src/masternodes/govvariables/lp_daily_dfi_reward.h:1]: It is 2020 already

File: src/masternodes/govvariables/lp_splits.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#include <core_io.h> /// ValueFromAmount
+#include <masternodes/masternodes.h> /// CCustomCSView
+#include <rpc/util.h> /// AmountFromValue
+
+
+Res LP_SPLITS::Import(const UniValue & val) {
+    if (!val.isObject())
+        return Res::Err("object of {poolId: rate,... } expected"); /// throw here? cause "AmountFromValue" can throw!
+    for (const std::string& key : val.getKeys()) {
+        auto id = DCT_ID::FromString(key);
```
Comment (ShengguangXiao) [inline src/masternodes/govvariables/lp_splits.cpp:16]: Use `const auto id`

File: src/masternodes/mn_checks.cpp [hunk 3]
```diff
@@ -466,15 +687,242 @@ Res ApplyAccountToAccountTx(CCustomCSView & mnview, CCoinsViewCache const & coin
     if (!res.ok) {
         return Res::ErrCode(CustomTxErrCodes::NotEnoughBalance, "%s: %s", base, res.msg);
     }
+
+    for (const auto& kv : SumAllTransfers(msg.to).balances) {
+        auto token = mnview.GetToken(kv.first);
```
Comment (ShengguangXiao) [inline src/masternodes/mn_checks.cpp:692]: use `const auto token`

File: src/masternodes/mn_rpc.cpp [hunk 4]
```diff
@@ -1230,88 +1356,239 @@ UniValue getaccount(const JSONRPCRequest& request) {
...
+               "The last optional argument (may be empty array) is an array of specific UTXOs to spend." +
                HelpRequiringPassphrase(pwallet) + "\n",
                {
-                    {"amounts", RPCArg::Type::OBJ, RPCArg::Optional::NO, "",
+                       {"from", RPCArg::Type::OBJ, RPCArg::Optional::NO, "",
                         {
-                            {"address", RPCArg::Type::STR, RPCArg::Optional::NO, "The defi address is the key, the value is amount in amount@token format. "
-                                                                                 "If multiple tokens are to be transferred, specify an array [\"amount1@t1\", \"amount2@t2\"]"},
+                                {"address", RPCArg::Type::STR, RPCArg::Optional::NO, "The defi address is the key, the value is amount in amount@token format. "
+                                                                                     "If multiple tokens are to be transferred, specify an array [\"amount1@t1\", \"amount2@t2\"]"},
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:1532]: The parameter description need to update

File: src/masternodes/mn_rpc.cpp [hunk 5]
```diff
@@ -1230,88 +1356,239 @@ UniValue getaccount(const JSONRPCRequest& request) {
...
-
-    // burn
-    const auto toBurn = SumAllTransfers(msg.to);
-    if (toBurn.balances.empty()) {
-        throw JSONRPCError(RPC_INVALID_PARAMETER, "zero amounts");
-    }
 
-    const auto txVersion = GetTransactionVersion(::ChainActive().Height());
-    CMutableTransaction rawTx(txVersion);
+    CMutableTransaction rawTx;
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:1579]: Create transaction by version

File: src/masternodes/mn_rpc.cpp [hunk 6]
```diff
@@ -1558,32 +2005,675 @@ UniValue accounttoutxos(const JSONRPCRequest& request) {
...
                RPCExamples{
-                       HelpExampleCli("listcommunitybalances", "")
-                       + HelpExampleRpc("listcommunitybalances", "")
+                       HelpExampleCli("createpoolpair",   "\"{\\\"tokenA\\\":\\\"MyToken1\\\","
+                                                          "\\\"tokenB\\\":\\\"MyToken2\\\","
+                                                          "\\\"commission\\\":\\\"0.001\\\","
+                                                          "\\\"status\\\":\\\"True\\\","
+                                                          "\\\"ownerAddress\\\":\\\"Address\\\""
+                                                          "}\" \"[{\\\"txid\\\":\\\"id\\\",\\\"vout\\\":0}]\" ")
+                       + HelpExampleRpc("createpoolpair", "\"{\\\"tokenA\\\":\\\"MyToken1\\\","
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2054]: These two examples are the same, should keep only one or update one for them.

File: src/masternodes/mn_rpc.cpp [hunk 7]
```diff
@@ -1558,32 +2005,675 @@ UniValue accounttoutxos(const JSONRPCRequest& request) {
...
+    metadata << static_cast<unsigned char>(CustomTxType::CreatePoolPair)
+             << poolPairMsg << pairSymbol;
+
+    CScript scriptMeta;
+    scriptMeta << OP_RETURN << ToByteVector(metadata);
+
+    CMutableTransaction rawTx;
+    rawTx.vout.push_back(CTxOut(0, scriptMeta));
+
+    for(std::set<CScript>::iterator it = Params().GetConsensus().foundationMembers.begin(); it != Params().GetConsensus().foundationMembers.end() && rawTx.vin.size() == 0; it++)
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2123]: Why not use `for (auto& it : Params().GetConsensus().foundationMembers)`

File: src/masternodes/mn_rpc.cpp [hunk 8]
```diff
@@ -1558,32 +2005,675 @@ UniValue accounttoutxos(const JSONRPCRequest& request) {
...
+                            },
+                        },
+                   },
+
+               },
+               RPCResult{
+                       "\"hash\"                  (string) The hex-encoded hash of broadcasted transaction\n"
+               },
+               RPCExamples{
+                       HelpExampleCli("updatetoken", "\"{\"pool\":\"POOL\",\"status\":true,"
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2188]: Change `updatetoken ` to `updatepoolpair`

File: src/masternodes/mn_rpc.cpp [hunk 9]
```diff
@@ -1558,32 +2005,675 @@ UniValue accounttoutxos(const JSONRPCRequest& request) {
...
+                                                     "\"commission\":0.01,\"ownerAddress\":\"Address\"}\" "
+                                                     "\"[{\"txid\":\"id\",\"vout\":0}]\"")
+                       + HelpExampleRpc("updatetoken", "\"{\"pool\":\"POOL\",\"status\":true,"
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2191]: Change `updatetoken ` to `updatepoolpair`

File: src/masternodes/mn_rpc.cpp [hunk 10]
```diff
@@ -1558,32 +2005,675 @@ UniValue accounttoutxos(const JSONRPCRequest& request) {
...
+    if (!metaObj["commission"].isNull()) {
+        commission = AmountFromValue(metaObj["commission"]);
+    }
+    if (!metaObj["ownerAddress"].isNull()) {
+        ownerAddress = DecodeScript(metaObj["ownerAddress"].getValStr());
+    }
+
+    CMutableTransaction rawTx;
+
+    for(std::set<CScript>::iterator it = Params().GetConsensus().foundationMembers.begin(); it != Params().GetConsensus().foundationMembers.end() && rawTx.vin.size() == 0; it++)
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2239]: Why not use for (auto& it : Params().GetConsensus().foundationMembers)

File: src/masternodes/mn_rpc.cpp [hunk 11]
```diff
@@ -1558,32 +2005,675 @@ UniValue accounttoutxos(const JSONRPCRequest& request) {
...
+            if (pool.totalLiquidity <= CPoolPair::MINIMUM_LIQUIDITY) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Pool is empty!");
+            }
+
+            auto resA = pool.reserveA;
+            auto resB = pool.reserveB;
+            if (poolSwapMsg.idTokenFrom != pool.idTokenA) {
+                std::swap(resA, resB);
+            }
+            arith_uint256 price256 = arith_uint256(resB) * CPoolPair::PRECISION / arith_uint256(resA);
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2393]: Before divide, need to check if resA is 0

File: src/masternodes/mn_rpc.cpp [hunk 12]
```diff
@@ -1558,32 +2005,675 @@ UniValue accounttoutxos(const JSONRPCRequest& request) {
...
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Current price +3% overflow!");
+            }
+            auto priceInt = price256 / CPoolPair::PRECISION;
+            poolSwapMsg.maxPrice.integer = priceInt.GetLow64();
+            poolSwapMsg.maxPrice.fraction = (price256 - priceInt * CPoolPair::PRECISION).GetLow64(); // cause there is no operator "%"
+
+            // price w/o +3% - cant overflow
+    //        price.integer = pool.reserveB / pool.reserveA;
+    //        price.fraction = (arith_uint256(pool.reserveB % pool.reserveA) * CPoolPair::PRECISION / arith_uint256(pool.reserveA)).GetLow64(); // sure < PRECISION
+        }
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2406]: Remove commented code

File: src/masternodes/mn_rpc.cpp [hunk 13]
```diff
@@ -1558,32 +2005,675 @@ UniValue accounttoutxos(const JSONRPCRequest& request) {
...
+    CWallet* const pwallet = GetWallet(request);
+
+    RPCHelpMan{"setgov",
+               "\nSet special parameters.\n",
+               {
+                    //{"govKey", RPCArg::Type::STR, RPCArg::Optional::NO,"First parameter - governance key."},
+                    {"parameters", RPCArg::Type::OBJ, RPCArg::Optional::NO, "",
+                        {
+                             {"name", RPCArg::Type::STR, RPCArg::Optional::NO,
+                              "Second parameter - governance data."},
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2572]: If the first parameter is commented, then this is the first parameter already

File: src/masternodes/mn_rpc.cpp [hunk 14]
```diff
@@ -1558,32 +2005,675 @@ UniValue accounttoutxos(const JSONRPCRequest& request) {
...
+    }
+    
+    CDataStream metadata(DfTxMarker, SER_NETWORK, PROTOCOL_VERSION);
+    metadata << static_cast<unsigned char>(CustomTxType::SetGovVariable)
+             << varStream;
+
+    CScript scriptMeta;
+    scriptMeta << OP_RETURN << ToByteVector(metadata);
+
+    CMutableTransaction rawTx;
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:2615]: Create transaction by version.

File: src/masternodes/poolpairs.cpp [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+        WriteBy<ByID>(WrapVarInt(poolID.v), pool);
+        return Res::Ok();
+    }
+    else if (poolPairByTokens && poolID != poolPairByTokens->first)
+    {
+        return Res::Err("Error, PoolID is incorrect");
+    }
+    else if (poolPairByTokens && (poolPairByTokens->second.idTokenA != pool.idTokenA || poolPairByTokens->second.idTokenB == pool.idTokenB))
+    {
+        throw std::runtime_error("Error, idTokenA or idTokenB is incorrect.");
```
Comment (ShengguangXiao) [inline src/masternodes/poolpairs.cpp:47]: This function other places return Res::Err, why throw here, where the exception is handled?

File: src/masternodes/poolpairs.cpp [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+        return Res::Err("Pool trading is turned off!");
+
+    bool const forward = in.nTokenId == idTokenA;
+
+    // it is important that reserves are at least SLOPE_SWAP_RATE (1000) to be able to slide, otherwise it can lead to underflow
+    if (reserveA < SLOPE_SWAP_RATE || reserveB < SLOPE_SWAP_RATE)
+        return Res::Err("Lack of liquidity.");
+
+    auto aReserveA = arith_uint256(reserveA);
+    auto aReserveB = arith_uint256(reserveB);
```
Comment (ShengguangXiao) [inline src/masternodes/poolpairs.cpp:118]: Add const

File: src/masternodes/poolpairs.cpp [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+
+    arith_uint256 maxPrice256 = arith_uint256(maxPrice.integer) * PRECISION + maxPrice.fraction;
+    arith_uint256 priceAB = (aReserveA * PRECISION / aReserveB);
+    arith_uint256 priceBA = (aReserveB * PRECISION / aReserveA);
```
Comment (ShengguangXiao) [inline src/masternodes/poolpairs.cpp:122]: Before divide need check aReserveA is 0?

File: src/masternodes/poolpairs.h [hunk 18]
```diff
@@ -0,0 +1,361 @@
...
+            if (!pool.swapEvent && (poolReward == 0 || pool.totalLiquidity == 0)) {
+                return true; // no events, skip to the next pool
+            }
+
+            ForEachPoolShare([&] (DCT_ID const & currentId, CScript const & provider) {
+                if (currentId != poolId) {
+                    return false; // stop
+                }
+                CAmount const liquidity = onGetBalance(provider, poolId).nValue;
+
```
Comment (ShengguangXiao) [inline src/masternodes/poolpairs.h:271]: Remove extra blank line

File: src/masternodes/mn_rpc.cpp (hunk 6 above)
Comment (DeFiDev) [reply src/masternodes/mn_rpc.cpp:2054]: Not exectly. One for "cli", one for "rpc" (HelpExampleCli, HelpExampleRpc) - it is a "common" practice in all bitcoin's rpcs.

Comment (DeFiDev) [reply src/masternodes/mn_rpc.cpp:2054]: I can just remove one, if you want

Comment (monstrobishi) [reply src/masternodes/mn_rpc.cpp:2054]: @DeFiDev that's ok, can keep both if they are for CLI and RPC

File: src/masternodes/mn_rpc.cpp (hunk 7 above)
Comment (DeFiDev) [reply src/masternodes/mn_rpc.cpp:2123]: Cause "&& rawTx.vin.size() == 0" there in 'for' clause. 
and it can't deduce proper type here on simple "auto&" or "auto const &". "auto" works, but it create (i think) an object copy here

Comment (DeFiDev) [reply src/masternodes/mn_rpc.cpp:2123]: I'll try to refactor it to improve readness later, but there are moorre places to refactor

File: src/masternodes/mn_rpc.cpp (hunk 10 above)
Comment (DeFiDev) [reply src/masternodes/mn_rpc.cpp:2239]: the same as previous

File: src/masternodes/mn_rpc.cpp (hunk 11 above)
Comment (DeFiDev) [reply src/masternodes/mn_rpc.cpp:2393]: They are can't be zero "by nature" if "pool.totalLiquidity > CPoolPair::MINIMUM_LIQUIDITY", even after trading

File: src/masternodes/poolpairs.cpp (hunk 15 above)
Comment (DeFiDev) [reply src/masternodes/poolpairs.cpp:47]: Nowhere. It is equal to assert here! This should not happen EVER, but for sure and easy debug by the first time.

File: src/masternodes/poolpairs.cpp (hunk 17 above)
Comment (DeFiDev) [reply src/masternodes/poolpairs.cpp:122]: No. See `if (reserveA < SLOPE_SWAP_RATE || reserveB < SLOPE_SWAP_RATE)` upthere

File: src/masternodes/mn_rpc.cpp (hunk 13 above)
Comment (DeFiDev) [reply src/masternodes/mn_rpc.cpp]: Thanks. Description was not correct in fact. See newer version

**Comment by DeFiDev** [comment]
All requested changes applyed, except those of commented by me separately

File: src/masternodes/mn_rpc.cpp (hunk 7 above)
Comment (ShengguangXiao) [reply src/masternodes/mn_rpc.cpp:2127]: Can put the checking `rawTx.vin.size() == 0` inside the loop, then it is easier to read, because this line is too long, at first glance i didn't see `rawTx.vin.size() == 0` is there.

Comment (DeFiDev) [reply src/masternodes/mn_rpc.cpp:2127]: Ok, but later. All "foundation auth" cases should be refactored at once, with distinct method

**Review by monstrobishi — APPROVED** [review]

**Comment by ShengguangXiao** [comment]
> 1. Includes all the latest liquidity functional:
>    Rpcs:
> 
> * createpoolpair
> * updatepoolpair
> * addpoolliquidity/removepoolliquidity
> * poolswap
>   Rpc listers:
> * listpoolpairs
> * getpoolpair
> * listpoolshares
> 
> 1. Includes 'government' feature. Rpcs:
> 
> * 'setgov' with two (for now) implemented "variables":
>   
>   * LP_DAILY_DFI_REWARD
>   * LP_SPLITS
> * getgov
> 
> 1. "totalSupply/minted" control, "finalizetoken" - those the very latest discussions - **not implemented.**
> 
> The very latest changes:
> 
> * modified "updatetoken" rpc (includes other token's flags) with new auth logic
> * modified "minttokens" rpc with new auth logic
> * new rpc "updatepoolpair" (changing status, trade commission and even owner)
> * all token's flags logic reviewed
> * pool trading on/off by pool "status" property
> 
> _Remark_:
> I recommend removing those
> `if (::ChainActive().Tip()->height < Params().GetConsensus().AMKHeight)`
> from rpcs so they are redundant and should be checked by first clauses **inside** "Apply*" methods (not in the main "switch" as for now). Besides the fact that now they are wrong in rpcs against current height (note, that all "apply's" called from rpc check the height as +1).
> Nothing of that was touched and wan not added ("AMKHeight ifs") in the new "pools" rpcs

If the tokenId is not used or initialized in the RPC handling function, then we can remove the tx version handling here. But in some functions, the tokenId should have value  for example `minttokens`.

**Comment by DeFiDev** [comment]
Update:
- That #57 PR (update tokens, mint DAT, etc) merged here.
- New fork height "bishanheight" implemented.
- All token's changes made were adjusted by this "ifs" with respect to consensus and 1.2.0 compatibility.
As part of the task, we ought to separate (make new version) of 'updatetoken' tx cause their serialization changed since 1.2.0. So, there will be two txs - one, that can only triggers "DAT" flag for pre-bishan height, and the second, that can change any tokens fields - after it. The RPC stays the same (it just makes different txs based on current height).

File: src/masternodes/mn_rpc.cpp [hunk 19]
```diff
@@ -918,12 +918,18 @@ UniValue tokenToJSON(DCT_ID const& id, CTokenImplementation const& token, bool v
         tokenObj.pushKV("isLPS", token.IsPoolShare());
         tokenObj.pushKV("finalized", token.IsFinalized());
 
-        tokenObj.pushKV("creationTx", token.creationTx.ToString());
-        tokenObj.pushKV("creationHeight", token.creationHeight);
-        tokenObj.pushKV("destructionTx", token.destructionTx.ToString());
-        tokenObj.pushKV("destructionHeight", token.destructionHeight);
         /// @todo tokens: collateral address/script
 //      tokenObj.pushKV("collateralAddress", token.destructionHeight);
+        if (id >= CTokensView::DCT_ID_START) {
```
Comment (DeFiDev) [inline src/masternodes/mn_rpc.cpp:923]: mismerged here

Comment (DeFiDev) [reply src/masternodes/mn_rpc.cpp:923]: don't need that "if" and casting to CTokenImplementation
all the difference (with applyed "minted" PR) - `tokenObj.pushKV("minted",` should appear

Comment (monstrobishi) [reply src/masternodes/mn_rpc.cpp:923]: @DeFiDev thanks for spotting that, can you fix this?

thank you

Comment (DeFiDev) [reply src/masternodes/mn_rpc.cpp:923]: ok

**Comment by DeFiDev** [comment]
@monstrobishi should I take care of conflicts? or you?

**Comment by monstrobishi** [comment]
@DeFiDev Hi. please take care of the conflicts, thank you.

**Comment by DeFiDev** [comment]
> @DeFiDev Hi. please take care of the conflicts, thank you.

done

**Comment by monstrobishi** [comment]
@DeFiDev can I ask for assistance with the conflicts again? it's the last time for this PR since we'll merge it soon 👍 

thank you

**Comment by DeFiDev** [comment]
> @DeFiDev can I ask for assistance with the conflicts again? it's the last time for this PR since we'll merge it soon +1

in progress... done already, but checking/testing

File: src/masternodes/govvariables/lp_daily_dfi_reward.cpp [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+UniValue LP_DAILY_DFI_REWARD::Export() const {
+    return ValueFromAmount(dailyReward);
+}
+
+Res LP_DAILY_DFI_REWARD::Validate(const CCustomCSView &) const
+{
+    // nothing to do
+    return Res::Ok();
```
Comment (ShengguangXiao) [inline src/masternodes/govvariables/lp_daily_dfi_reward.cpp:23]: Why don't need Validate here?

File: src/arith_uint256.cpp [hunk 21]
```diff
@@ -184,6 +184,29 @@ unsigned int base_uint<BITS>::bits() const
...
+    while (bit != 0) {
+        if (num >= res + bit) {
+            num -= res + bit;
+            res = (res >> 1) + bit;
+        } else
+            res >>= 1;
+        bit >>= 2;
+    }
+    return res;
+}
```
Comment (ShengguangXiao) [inline src/arith_uint256.cpp:208]: Can add some unit test cases for this function?

**Review by ShengguangXiao — APPROVED** [review]

File: src/test/liquidity_tests.cpp [hunk 22]
```diff
@@ -0,0 +1,377 @@
...
+                    BOOST_CHECK(rwdB == id.v * COIN * 2 / ProvidersCount);
+                }
+                return true;
+            }, PoolShareKey{idPool, {}});
+        }
+    }
+}
+
+
+BOOST_AUTO_TEST_SUITE_END()
```
Comment (ShengguangXiao) [inline src/test/liquidity_tests.cpp:377]: This test only covered AddLiquidity feature, but haven't cover the RemoveLiquidity feature.

File: src/masternodes/govvariables/lp_daily_dfi_reward.cpp (hunk 20 above)
Comment (DeFiDev) [reply src/masternodes/govvariables/lp_daily_dfi_reward.cpp:23]: For example? I have no any criteria to validate "daily dfi reward". <0? >MAX_MONEY? 
It comes from "foundation" only, and firstly checked with auth on apply. Basic ranges was checked in import (called from rpc) with AmountFromValue.
One of foundation members will cheat, manually constructing tx bytecode to avoid rpc's AmountFromValue to place into tx negative value?

File: src/test/liquidity_tests.cpp (hunk 22 above)
Comment (DeFiDev) [reply src/test/liquidity_tests.cpp:377]: This test is exactly for testing math. Math with boundary values (too large or too low) that was hard to simulate by functional python's rpc calls. Or, math of rewards (which is hard to set by rpc's too). 
"removepoolliquidity", from this point of view, is trivial and tested by "feature_poolpair_liquidity.py"

File: src/masternodes/govvariables/lp_daily_dfi_reward.cpp (hunk 20 above)
Comment (DeFiDev) [reply src/masternodes/govvariables/lp_daily_dfi_reward.cpp:23]: Another words, the general validation of "setgov" is an auth. All others are only aimed at protecting "the gods" from hurtful mistakes (as with total sum of 100% in LP_SPLITS, or pool existence)

File: src/arith_uint256.cpp (hunk 21 above)
Comment (DeFiDev) [reply src/arith_uint256.cpp:208]: added, see last commit

**Comment by ShengguangXiao** [comment]
@DeFiDev The `createpoolpair` second parameter should be optional,  just like `createtoken`, but if don't input the second parameter, it will report below error. Need to do same handling as `createtoken`.
```
error code: -1
error message:
JSON value is not an array as expected
```

**Review by ShengguangXiao — APPROVED** [review]

**Comment by ShengguangXiao** [comment]
> @DeFiDev The `createpoolpair` second parameter should be optional, just like `createtoken`, but if don't input the second parameter, it will report below error. Need to do same handling as `createtoken`.
> 
> ```
> error code: -1
> error message:
> JSON value is not an array as expected
> ```

This has been fixed my PR https://github.com/DeFiCh/ain/pull/65  already

**Comment by ShengguangXiao** [comment]
@DeFiDev In mn_rcp.cpp, the `const auto txVersion = GetTransactionVersion(targetHeight);` is using current height + 1 as target height, then the below get version also need to follow in in `wallet.cpp`

```C++
const auto txVersion = GetTransactionVersion(locked_chain.getHeight().get_value_or(0));
```
